exchanges/coinbasepro/coinbasepro_test.go

```diff
@@ -2,1064 +2,1981 @@ package coinbasepro
+	p, err := c.GetEnabledPairs(asset.Spot)
+	require.NoError(t, err)
+	exp := subscription.List{}
+	for _, baseSub := range defaultSubscriptions.Enabled() {
+		s := baseSub.Clone()
+		s.QualifiedChannel = subscriptionNames[s.Channel]
+		if s.Asset != asset.Empty {
+			s.Pairs = p
+		}
+		exp = append(exp, s)
 	}
+	subs, err := c.generateSubscriptions()
+	require.NoError(t, err)
+	testsubs.EqualLists(t, exp, subs)
+
+	_, err = subscription.List{{Channel: "wibble"}}.ExpandTemplates(c)
+	assert.ErrorContains(t, err, "subscription channel not supported: wibble")
 }
 
-func TestWsOrderbook(t *testing.T) {
-	pressXToJSON := []byte(`{
-    "type": "snapshot",
-    "product_id": "BTC-USD",
-    "bids": [["10101.10", "0.45054140"]],
-    "asks": [["10102.55", "0.57753524"]],
-	"time":"2023-08-15T06:46:55.376250Z"
-}`)
-	err := c.wsHandleData(pressXToJSON)
-	if err != nil {
-		t.Error(err)
-	}
+func TestSubscribeUnsubscribe(t *testing.T) {
+	t.Parallel()
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, c)
+	req := subscription.List{{Channel: "heartbeat", Asset: asset.Spot,
+		Pairs: currency.Pairs{currency.NewPairWithDelimiter(testCrypto.String(), testFiat.String(), "-")}}}
+	err := c.Subscribe(req)
+	assert.NoError(t, err)
+	err = c.Unsubscribe(req)
+	assert.NoError(t, err)
+}
 
-	pressXToJSON = []byte(`{
-  "type": "l2update",
-  "product_id": "BTC-USD",
-  "time": "2023-08-15T06:46:57.933713Z",
-  "changes": [
-    [
-      "buy",
-      "10101.80000000",
-      "0.162567"
-    ]
-  ]
-}`)
-	err = c.wsHandleData(pressXToJSON)
-	if err != nil {
-		t.Error(err)
+func TestCheckSubscriptions(t *testing.T) {
+	t.Parallel()
+	c := &CoinbasePro{ //nolint:govet // Intentional shadow to avoid future copy/paste mistakes
+		Base: exchange.Base{
+			Config: &config.Exchange{
+				Features: &config.FeaturesConfig{
+					Subscriptions: subscription.List{
+						{Enabled: true, Channel: "matches"},
+					},
+				},
+			},
+			Features: exchange.Features{},
+		},
 	}
+	c.checkSubscriptions()
+	testsubs.EqualLists(t, defaultSubscriptions.Enabled(), c.Features.Subscriptions)
+	testsubs.EqualLists(t, defaultSubscriptions, c.Config.Features.Subscriptions)
 }
 
-func TestWsOrders(t *testing.T) {
-	pressXToJSON := []byte(`{
-    "type": "received",
-    "time": "2014-11-07T08:19:27.028459Z",
-    "product_id": "BTC-USD",
-    "sequence": 10,
-    "order_id": "d50ec984-77a8-460a-b958-66f114b0de9b",
-    "size": "1.34",
-    "price": "502.1",
-    "side": "buy",
-    "order_type": "limit"
-}`)
-	err := c.wsHandleData(pressXToJSON)
-	if err != nil {
-		t.Error(err)
+func TestGetJWT(t *testing.T) {
+	t.Parallel()
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, c)
+	creds, err := c.GetCredentials(context.Background())
+	assert.NoError(t, err)
+	_, err = c.GetJWT(context.Background(), "")
+	if strings.HasPrefix(creds.Secret, "-----BEGIN EC PRIVATE KEY-----\n") {
+		assert.NoError(t, err)
+	} else {
+		assert.ErrorIs(t, err, errCantDecodePrivKey)
```

Comment:
I don't see that as an acceptable pass and it is okay to fail while you don't support it

Could you please add todo commentary that this function doesn't work? You've added a todo elsewhere talking about supporting the new api key format, but the function to utilise it doesn't work even when set

Reply:
It should work, in that it'd generate a valid JWT and be usable if the auth functions are tweaked a little, it's just not called.

Where would you want that commentary set; in the test, or the function itself?